src/writer.rs

```diff
@@ -146,70 +179,304 @@ impl<'a, W: Write> Writer<'a, W> {
         Ok(num_bytes)
     }
 
+    /// Return what the `Writer` is writing to, consuming the `Writer` itself.
+    ///
+    /// **NOTE** This function doesn't guarantee that everything gets written before consuming the
+    /// buffer. Please call [`flush`](struct.Writer.html#method.flush) before.
     pub fn into_inner(self) -> W {
         self.writer
     }
 }
 
-pub struct SingleWriter<'a> {
-    schema: &'a Schema,
-    header: Vec<u8>,
-    buffer: Vec<u8>,
-    codec: Codec,
-}
-
-impl<'a> SingleWriter<'a> {
-    pub fn new(schema: &'a Schema) -> SingleWriter<'a> {
-        Self::with_codec(schema, Codec::Null)
-    }
+/// Create an Avro header given a schema, a codec and a sync marker.
+fn header(schema: &Schema, codec: Codec, marker: &[u8]) -> Result<Vec<u8>, Error> {
```

Comment:
nit: Now that `SingleWriter`, this is only used in the `Writer` trait, and is private. Would it make sense to move it inside the implementation of `Writer`?

Or perhaps keeping it like this would make testing easier? 🤷‍♂️

Reply:
I think it makes sense, since the whole codebase is using methods a lot.